autode/opt/coordinates/dic.py

```diff
@@ -199,42 +193,277 @@ def iadd(self,
+
+            # and the d^2L/ds_i dλ_i = -dC_i/ds_i = -1
+            #         d^2L/dλ_i dλ_j = 0
+
+            for i in range(m):
+                self.h[n+i, :] = self.h[:, n+i] = 0.
+
+            for i in range(m):
+                self.h[n-m+i, n+i] = self.h[n+i, n-m+i] = -1.
+
+        return None
+
+    def h_or_h_inv_has_correct_shape(self, arr: Optional[np.ndarray]):
+        """Does a Hessian or its inverse have the correct shape?"""
+        if arr is None:
+            return True  # None is always valid
+
+        n_rows, n_cols = arr.shape
+        return arr.ndim == 2 and n_rows == n_cols == len(self) + self.n_constraints
+
+    def update_lagrange_multipliers(self, arr: np.ndarray) -> None:
+        """Update the lagrange multipliers by adding a set of values"""
+
+        if arr.shape != self._lambda.shape:
+            raise ValueError("Cannot set lagrange multipliers. Incorrect shape")
+
+        self._lambda[:] = np.asarray(self._lambda) + np.asarray(arr)
+        return None 
+    
+
+def _schmidt_orthogonalise(arr: np.ndarray, *indexes: int) -> np.ndarray:
+    """
+    Perform Schmidt orthogonalization to generate orthogonal vectors
+    that include a number of unit vectors, the non-zero components of which
+    are defined by indexes. This generates a transform matrix U which will
+    provide pure primitive coordinates, which can then be constrained simply
+    """
+    logger.info(f"Schmidt-orthogonalizing. Using {indexes} as "
+                f"orthonormal vectors")
+
+    u = np.zeros_like(arr)
+    _, n = arr.shape
+    m = len(indexes)
+
+    # Set the unit vectors as the first m columns
+    for i, index in enumerate(indexes):
+        u[index, i] = 1.
+
+    # and the remaining n-m columns as the orthogonalised values
+    for i in range(m, n):
+
+        u_i = arr[:, i]
+        for j in range(0, i):
+            u_i -= proj(u[:, j], arr[:, i])
+
+        u_i /= np.linalg.norm(u_i)
+
+        u[:, i] = u_i.copy()
+
+    # Arbitrarily place the defined unit vectors at the end
+    permutation = list(range(m, n)) + list(range(m))
+    return u[:, permutation]
+
+
+def _symmetry_inequivalent_u(u, q) -> np.ndarray:
+    """Remove symmetry equivalent vectors from the U matrix"""
+
+    # The non-redundant space can be further pruned by considering symmetry
+    idxs = []
+    s = np.matmul(u.T, q)
+
+    for i, s_i in enumerate(s):
+        is_unique = all(not np.isclose(s_i, s[j], atol=1E-20) for j in idxs)
+
+        if is_unique or _is_pure_primitive(u[:, i]):
+            idxs.append(i)
+
+    logger.info(f"Removing {len(s) - len(idxs)} symmetry equiv. DICs")
+    return u[:, idxs]
+
+
+def _is_pure_primitive(v: np.ndarray) -> bool:
+    """
+    Is this vector a pure primitive? where all but one of the coefficients
+    are zero.
+    """
+
+    def n_values_close_to(value):
+        return sum(np.isclose(v_i, value) for v_i in v)
```

Comment:
Probably fine, but does an explicit tolerance need to be defined if values often differ by > the default tolerance?

Reply:
clearer if there is one definitely – will add 👍🏼